src/librustc_middle/mir/mod.rs

```diff
@@ -88,6 +88,19 @@ impl MirPhase {
     }
 }
 
+/// Coverage data computed by the `InstrumentCoverage` MIR pass, when compiling with
+/// `-Zinstrument_coverage`.
+#[derive(Clone, RustcEncodable, RustcDecodable, Debug, HashStable, TypeFoldable)]
+pub struct CoverageData {
+    /// A hash value that can be used by the consumer of the coverage profile data to detect
+    /// changes to the instrumented source of the associated MIR body (typically, for an
+    /// individual function).
+    pub hash: u64,
+
+    /// The total number of coverage region counters added to this MIR Body.
+    pub num_counters: usize,
```

Comment:
Agreed. I tried u32 first, but for some reason u64 and usize are Foldable, and u32 is not; and I couldn't figure out where that's getting set. I'll look more closely.

Reply:
I'm able to get it to work by explicitly implementing `TypeFoldable` for `u32`. Will upload for review shortly.

Reply:
Oh, looking at your last comment more closely, if I don't have to add coverage data to the body, I may not need it to be FoldableType. Hmm...

Reply:
Seems fine to add an impl for it (I think this is the correct place)

https://github.com/rust-lang/rust/blob/a39c7787ba246353178e099373b9240be0d9e603/src/librustc_middle/ty/structural_impls.rs#L260

but if we can get away without it, that's even better.

Reply:
FYI, I had put it here:

https://github.com/rust-lang/rust/blob/a39c7787ba246353178e099373b9240be0d9e603/src/librustc_middle/mir/type_foldable.rs#L322

I'm not sure it matters, but since it's a primitive (not related to `ty`) I thought keeping the impl with the type definition made sense. It's hard to tell why some impls are where they are for this trait.

Reply:
I went ahead and committed these changes while I'm waiting for any additional guidance you can provide on your other suggested change to implement coverage_data as a query.